### PR TITLE
WIP: Added integer option to slider control, added slope editor widget

### PR DESCRIPTION
**Pull request by BAndiT1983**

### CODE REVIEW

**Comment by BAndiT1983** [comment]
image
image
image

**Comment by se6astian-pi** [comment]
1 slope = HDR PLR off
no kneepoint
just standard exposure

2 slopes means kneepoint 1 and vtfl1 are active

3 slopes means kneepoint 1 &2 and vtftl 1& 2 are active

Also the 0-100% range of choices was created for the kneepoint level in my dashboard not the exposure time.

File: frontend/components/NctrlValueWidgets.jsx [hunk 1]
```diff
@@ -233,3 +257,97 @@ export function NctrlValueText({ text }) {
...
+    setSelectedIndex(index);
+  };
+
+  const showSlopes2 = selectedIndex >= 1 ? '' : 'none';
+  const showSlopes3 = selectedIndex == 2 ? '' : 'none';
+  const showMessage = selectedIndex == 0 ? '' : 'none';
+
+  return (
+    <div style={{ display: hide }}>
+      {NctrlValueButtons({
```
Comment (anuejn) [inline frontend/components/NctrlValueWidgets.jsx]: why not use the jsx syntax here?

File: frontend/components/NctrlValueWidgets.jsx [hunk 2]
```diff
@@ -233,3 +257,97 @@ export function NctrlValueText({ text }) {
...
+
+  const [selectedIndex, setSelectedIndex] = useState('');
+
+  const getSelectedIndex = index => {
+    setSelectedIndex(index);
+  };
+
+  const showSlopes2 = selectedIndex >= 1 ? '' : 'none';
+  const showSlopes3 = selectedIndex == 2 ? '' : 'none';
+  const showMessage = selectedIndex == 0 ? '' : 'none';
```
Comment (anuejn) [inline frontend/components/NctrlValueWidgets.jsx]: probably you want to have 'block' as the other option

File: frontend/components/NctrlValueWidgets.jsx [hunk 3]
```diff
@@ -233,3 +257,97 @@ export function NctrlValueText({ text }) {
...
   return <Typography className={classes.text}>{text}</Typography>;
 }
+
+export function NctrlValueSlopeeditor({ text, hidden, rerender, rerenderDeps }) {
+  const classes = useStyles();
+  const hide = hidden ? 'none' : '';
+
+  const [selectedIndex, setSelectedIndex] = useState('');
+
+  const getSelectedIndex = index => {
```
Comment (anuejn) [inline frontend/components/NctrlValueWidgets.jsx]: this simply does nothing?

File: frontend/components/NctrlValueWidgets.jsx [hunk 4]
```diff
@@ -233,3 +257,97 @@ export function NctrlValueText({ text }) {
   const classes = useStyles();
   return <Typography className={classes.text}>{text}</Typography>;
 }
+
+export function NctrlValueSlopeeditor({ text, hidden, rerender, rerenderDeps }) {
+  const classes = useStyles();
+  const hide = hidden ? 'none' : '';
```
Comment (anuejn) [inline frontend/components/NctrlValueWidgets.jsx]: what is this prop good for?

File: frontend/components/NctrlValueWidgets.jsx [hunk 5]
```diff
@@ -111,13 +119,15 @@ export function NctrlValueTextfield({ path, rerender, rerenderDep }) {
...
+export function NctrlValueSlider({ path, options, min, max, integer, rerender, rerenderDep }) {
   const classes = useStyles();
 
+  const parseValue = integer ? parseInt : parseFloat;
+
   const nctrlValue = NctrlValue.of(path);
   const [value, setValue] = useState(0.0);
   useEffect(() => {
-    nctrlValue.value().then(x => setValue(parseFloat(x)));
+    nctrlValue.value().then(x => setValue(parseValue(x)));
```
Comment (anuejn) [inline frontend/components/NctrlValueWidgets.jsx]: did you test it? does it work?

File: frontend/components/NctrlValueWidgets.jsx [hunk 6]
```diff
@@ -143,13 +155,13 @@ export function NctrlValueSlider({ path, options, min, max, rerender, rerenderDe
     } else if (value < min) {
       return <ArrowLeft />;
     } else {
-      return value.toFixed(value > 100 ? 0 : 1);
+      return value.toFixed(integer || value > 100 ? 0 : 1);
```
Comment (anuejn) [inline frontend/components/NctrlValueWidgets.jsx]: what does this do?

File: frontend/components/NctrlValueWidgets.jsx [hunk 7]
```diff
@@ -193,29 +205,41 @@ export function NctrlValueSlider({ path, options, min, max, rerender, rerenderDe
...
+  // Set first button as active by default
+  const [active, setActive] = useState(Object.keys(buttons)[0]);
 
   // the buttons are a map of label -> (formerValue) => nextValue
   return (
     <ButtonGroup size="large" className={classes.input}>
-      {Object.keys(buttons).map(label => {
+      {Object.keys(buttons).map((label, index) => {
         const clickFn = eval(buttons[label]);
+        const activeClass = active === label ? classes.active : '';
```
Comment (anuejn) [inline frontend/components/NctrlValueWidgets.jsx]: is this really the way to do this?

**Review by anuejn — COMMENTED** [review]
Hey, nice you picked it up, I left some comments :)

File: frontend/components/NctrlValueWidgets.jsx (hunk 4 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: This was left from a test regarding hiding components, not required at the moment and will be ditched.

File: frontend/components/NctrlValueWidgets.jsx (hunk 3 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: Set 0 as default now.

File: frontend/components/NctrlValueWidgets.jsx (hunk 6 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: Ensures that label shows integer values when the slider options integer is enabled.

File: frontend/components/NctrlValueWidgets.jsx (hunk 7 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: Have followed some docs and tutorials on React which were manipulating the classes. Have to check if there is a better way.

File: frontend/components/NctrlValueWidgets.jsx (hunk 2 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: Done.

File: frontend/components/NctrlValueWidgets.jsx (hunk 1 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: Obsolete attribute, but nevertheless, what would be the JSX syntax?

File: frontend/components/NctrlValueWidgets.jsx (hunk 6 above)
Comment (anuejn) [reply frontend/components/NctrlValueWidgets.jsx]: i see

File: frontend/components/NctrlValueWidgets.jsx (hunk 5 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: Was working, but could not ensure properly, so reverted to parseFloat, seems that the value is already converted by toFixed(0) which results in "integer".

File: frontend/components/NctrlValueWidgets.jsx (hunk 1 above)
Comment (anuejn) [reply frontend/components/NctrlValueWidgets.jsx]: I guess you can Just write `<NctrlValueButtons prop={value} />`

File: frontend/components/NctrlValueWidgets.jsx (hunk 3 above)
Comment (anuejn) [reply frontend/components/NctrlValueWidgets.jsx]: ?

File: frontend/components/NctrlValueWidgets.jsx (hunk 7 above)
Comment (anuejn) [reply frontend/components/NctrlValueWidgets.jsx]: couldnt you just pass disabeled to the button? or is there something else going on here?

Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: It's not disabled, just set active there.

Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: Maybe toggle buttons are the better choice there -> https://mui.com/components/toggle-button/

File: frontend/components/NctrlValueWidgets.jsx (hunk 3 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: please explain your comment, as this code is providing me with the index of active button.

File: frontend/components/NctrlValueWidgets.jsx (hunk 7 above)
Comment (anuejn) [reply frontend/components/NctrlValueWidgets.jsx]: yup sounds like it

File: frontend/components/NctrlValueWidgets.jsx (hunk 3 above)
Comment (anuejn) [reply frontend/components/NctrlValueWidgets.jsx]: you are wrapping a function with another function simply passing through its argument. so the wrapper doesnt add any value and you can just call the inner function directly

Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: ah, ok, wil check a bit later, thanks for the hint

File: frontend/components/NctrlValueWidgets.jsx (hunk 4 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: removed

File: frontend/components/NctrlValueWidgets.jsx (hunk 3 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: is now replaced with inline for toggle buttons

File: frontend/components/NctrlValueWidgets.jsx (hunk 1 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: will check which properties can be set directly, was referring to the usual HTML/CSS way there

**Comment by BAndiT1983** [comment]
image

File: frontend/components/NctrlValueWidgets.jsx [hunk 8]
```diff
@@ -127,6 +139,8 @@ export function NctrlValueSlider({ path, options, min, max, rerender, rerenderDe
     const currentTime = +new Date();
     if (lastUpdate + MIN_SEND_DELAY < currentTime && newValue !== value) {
       setLastUpdate(currentTime);
+      console.log('VALUE: ' + newValue);
```
Comment (anuejn) [inline frontend/components/NctrlValueWidgets.jsx]: :printer:

**Review by anuejn — APPROVED** [review]
fine for me :)

**Comment by anuejn** [comment]
I didnt test this thoug - maybe @sp-apertus can give it a shot :)

File: frontend/components/NctrlValueWidgets.jsx (hunk 8 above)
Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: Ah, right, forgot to remove this one

Comment (BAndiT1983) [reply frontend/components/NctrlValueWidgets.jsx]: done

**Comment by se6astian-pi** [comment]
Sure, I can test it!

**Comment by se6astian-pi** [comment]
I just tried and get a white/empty dashboard.

Are you sure that there is nothing missing as there is only one hit of "slopeeditor" with the changes...

**Comment by se6astian-pi** [comment]
I was able to preview it now, very nice!

One odd behaviour though was noticed when clikcing an already active toggle button.
It goes disabled and then no PLR option appears active.


Also the PLR GUI does not seem to actually set any values, at least the register explorer shows no changes.

**Comment by BAndiT1983** [comment]
Active button problem is fixed, regarding the parameters, we need to check the paths and if the values arrive in daemon. Maybe a quick pairing session will do.

**Comment by rroohhh** [comment]
You can also run the daemon locally in a mocked mode to test it: https://github.com/apertus-open-source-cinema/nctrl#developing-locally

**Comment by se6astian-pi** [comment]
active slope button issue confirmed fixed.
Setting sensor parameters not working at all, throws console errors:
https://pastebin.com/35Qs1ZCV

**Pull request merged**